cleanrl/ppo_dna_atari_envpool.py

```diff
@@ -0,0 +1,468 @@
+
+    agent_policy = Agent(envs).to(device)
+    agent_value = Agent(envs).to(device)
+    policy_optimizer = optim.Adam(agent_policy.parameters(), lr=args.policy_learning_rate, eps=1e-5)
+    value_optimizer = optim.Adam(agent_value.parameters(), lr=args.value_learning_rate, eps=1e-5)
+    distill_optimizer = optim.Adam(agent_policy.parameters(), lr=args.distill_learning_rate, eps=1e-5)
+
+    if args.track:
+        # log gradients
+        wandb.watch(agent_policy)
+        wandb.watch(agent_value)
+
+    # ALGO Logic: Storage setup
+    obs = torch.zeros((args.num_steps, args.num_envs) + envs.single_observation_space.shape).to(device)
+    action_logits = torch.zeros((args.num_steps, args.num_envs, envs.single_action_space.n)).to(device)
+    actions = torch.zeros((args.num_steps, args.num_envs) + envs.single_action_space.shape).to(device)
+    logprobs = torch.zeros((args.num_steps, args.num_envs)).to(device)
+    rewards = torch.zeros((args.num_steps, args.num_envs)).to(device)
+    dones = torch.zeros((args.num_steps, args.num_envs)).to(device)
+    values = torch.zeros((args.num_steps, args.num_envs)).to(device)
+    avg_returns = deque(maxlen=20)
+
+    # TRY NOT TO MODIFY: start the game
+    global_step = 0
+    env_step = 0
+    start_time = time.time()
+    next_obs = torch.tensor(envs.reset(), dtype=torch.float32).to(device)
+    next_done = torch.zeros(args.num_envs).to(device)
+    num_updates = args.total_timesteps // args.epoch_size
+
+    for update in range(1, num_updates + 1):
+        # Annealing the rate if instructed to do so.
+        if args.anneal_lr:
+            frac = 1.0 - (update - 1.0) / num_updates
+            policy_optimizer.param_groups[0]["lr"] = frac * args.policy_learning_rate
+            value_optimizer.param_groups[0]["lr"] = frac * args.value_learning_rate
+            distill_optimizer.param_groups[0]["lr"] = frac * args.distill_learning_rate
+
+        for step in range(0, args.num_steps):
+            global_step += 1 * args.num_envs
+            env_step += envs.spec.config.frame_skip * args.num_envs
+            obs[step] = next_obs
+            dones[step] = next_done
+
+            # ALGO LOGIC: action logic
+            with torch.no_grad():
+                logit, _ = agent_policy(next_obs)
+                _, value = agent_value(next_obs)
+                action, logprob = Agent.sample_action(logit)
+            action_logits[step] = logit
+            actions[step] = action
+            logprobs[step] = logprob
+            values[step] = value
+
+            # TRY NOT TO MODIFY: execute the game and log data.
+            next_obs, reward, done, info = envs.step(action.cpu().numpy())
+            rewards[step] = torch.tensor(reward, dtype=torch.float32).to(device).view(-1)
+            next_obs = torch.tensor(next_obs, dtype=torch.float32).to(device)
+            next_done = torch.tensor(done, dtype=torch.float32).to(device)
```

Comment:
This difference should be kept to a minimum.

<img width="1469" alt="image" src="https://user-images.githubusercontent.com/5555347/180087529-2a001df2-92c0-485f-969b-90f20c5ca01e.png">

Reply:
done